collector/src/main/java/io/prometheus/jmx/JmxCollector.java

```diff
@@ -368,99 +380,134 @@ public void recordBean(
         String help = attrDescription + " (" + beanName + attrName + ")";
         String attrNameSnakeCase = toSnakeAndLowerCase(attrName);
 
-        for (Rule rule : config.rules) {
-          Matcher matcher = null;
-          String matchName = beanName + (rule.attrNameSnakeCase ? attrNameSnakeCase : attrName);
-          if (rule.pattern != null) {
-            matcher = rule.pattern.matcher(matchName + ": " + beanValue);
-            if (!matcher.matches()) {
-              continue;
+        MatchedRule matchedRule = null;
+        String cacheName = beanName + attrName;
+        if (config.cacheRules) {
+          matchedRule = cachedRules.get(cacheName);
+        }
+
+        if (matchedRule == null) {
+          for (Rule rule : config.rules) {
+            Matcher matcher = null;
+            String matchName = beanName + (rule.attrNameSnakeCase ? attrNameSnakeCase : attrName);
+
+            // Using bean value in caching is not possible (caching is only done on bean name, values can change)
+            if (!config.cacheRules) {
```

Comment:
This creates two subtly different formats to match against, which is going to cause confusion.